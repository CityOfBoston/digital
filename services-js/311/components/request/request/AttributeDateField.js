// @flow

import React from 'react';
import { css } from 'glamor';
import { action, observable, reaction, computed } from 'mobx';
import { observer } from 'mobx-react';

import type Question from '../../../data/store/Question';

export type Props = {|
  question: Question,
|};

const INPUT_STYLE = css({
  textAlign: 'right',
  width: '10em',
  marginRight: '1em',
});

const INPUT_MESSAGE_CONTAINER_STYLE = css({
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
});

function maybeRenderRequired(required: boolean) {
  if (required) {
    return <span className="t--req">Required</span>;
  } else {
    return null;
  }
}

@observer
export default class AttributeDateField extends React.Component<Props> {
  @observable dateValue: string = '';
  @observable timeValue: string = '';

  @observable dateFocus: boolean = false;
  @observable timeFocus: boolean = false;

  formUpdaterDisposer: Function;
  valueWritebackDisposer: Function;

  componentDidMount() {
    // Updates the form based on the question value
    this.formUpdaterDisposer = reaction(
      () => this.questionDate,
      (date: ?Date) => {
        if (date) {
          this.dateValue = `${date.getMonth() +
            1}/${date.getDate()}/${date.getFullYear()}`;

          const hours = date.getHours();
          const minutes = date.getMinutes();

          const h = hours > 12 ? hours - 12 : hours;
          const m = minutes < 10 ? `0${minutes}` : minutes;

          const ampm = hours >= 12 && hours < 24 ? 'pm' : 'am';
          this.timeValue = `${h === 0 ? 12 : h}:${m} ${ampm}`;
        } else {
          this.dateValue = '';
          this.timeValue = '';
        }
      },
      {
        fireImmediately: true,
        name: 'date form updater',
      }
    );

    // when the valid form value changes, writes it back to the question. should
    // automatically stablize with the above updater.
    this.valueWritebackDisposer = reaction(
      () => this.validValue,
      (validValue: ?string) => {
        const { question } = this.props;

        if (validValue === null) {
          question.malformed = true;
        } else {
          question.malformed = false;
          question.value = validValue;
        }
      },
      {
        name: 'date value writeback',
      }
    );
  }

  componentWillUnmount() {
    this.formUpdaterDisposer();
    this.valueWritebackDisposer();
  }

  // returns a Date value for the current question.
  @computed
  get questionDate(): ?Date {
    const { question: { value } } = this.props;
    if (value) {
      if (value.indexOf('T') >= 0) {
        return new Date(value);
      } else {
        // In this case we have a date value but no time. We want to convert
        // it into a Date value for that day in the current timezone.
        const [year, month, day] = value.split('-').map(v => parseInt(v, 10));
        return new Date(year, month - 1, day);
      }
    } else {
      return null;
    }
  }

  @action.bound
  handleDateChange(ev: SyntheticInputEvent<>) {
    this.dateValue = ev.target.value;
  }

  @action.bound
  handleTimeChange(ev: SyntheticInputEvent<>) {
    this.timeValue = ev.target.value;
  }

  @action.bound
  handleDateFocus() {
    this.dateFocus = true;
  }

  @action.bound
  handleDateBlur() {
    this.dateFocus = false;
  }

  @action.bound
  handleTimeFocus() {
    this.timeFocus = true;
  }

  @action.bound
  handleTimeBlur() {
    this.timeFocus = false;
  }

  @computed
  get showTime(): boolean {
    return this.props.question.type === 'DATETIME';
  }

  @computed
  get datePieces(): ?Object {
    const datePieces = (this.dateValue || '').split('/');
    if (datePieces.length !== 3) {
      return null;
    }

    const month = parseInt(datePieces[0], 10);
    const day = parseInt(datePieces[1], 10);
    const year = parseInt(datePieces[2], 10);

    if (!(month && month >= 1 && month <= 12)) {
      return null;
    }

    if (!(day && day >= 1 && day <= 31)) {
      return null;
    }

    // y3k
    if (!(year && year >= 1900 && year < 3000)) {
      return null;
    }

    return { month, day, year };
  }

  @computed
  get timePieces(): ?Object {
    const timeMatch = (this.timeValue || '')
      .match(/(\d+):(\d\d)\s*([aApP][mM])/);
    if (!timeMatch) {
      return null;
    }

    let hours = parseInt(timeMatch[1], 10);
    const minutes = parseInt(timeMatch[2], 10);

    if (timeMatch[3] && timeMatch[3][0].toLowerCase() === 'p' && hours !== 12) {
      hours += 12;
    }

    if (timeMatch[3] && timeMatch[3][0].toLowerCase() === 'a' && hours === 12) {
      hours -= 12;
    }

    if (hours === 24) {
      hours = 0;
    }

    if (hours < 0 || hours > 23) {
      return null;
    }

    if (minutes < 0 || minutes > 59) {
      return null;
    }

    return { hours, minutes };
  }

  @computed
  get isDateValid(): boolean {
    return (
      (this.dateValue === '' && !this.timeValue) || this.datePieces !== null
    );
  }

  @computed
  get isTimeValid(): boolean {
    return (
      !this.showTime ||
      (this.timeValue === '' && !this.dateValue) ||
      this.timePieces !== null
    );
  }

  /* combines date and time if necessary into an ISO string to write to
     the store. Returns null if the current values are not parsable, and '' if
     they're both empty.
  */
  @computed
  get validValue(): ?string {
    if (!this.dateValue && !this.timeValue) {
      return '';
    }

    const datePieces = this.datePieces;
    if (!datePieces) {
      return null;
    }

    const { month, day, year } = datePieces;

    if (this.showTime) {
      const timePieces = this.timePieces;

      if (!timePieces) {
        return null;
      }

      const { hours, minutes } = timePieces;

      const date = new Date(year, month - 1, day, hours, minutes);
      return date.toISOString();
    } else {
      return `${year}-${month < 10 ? `0${month}` : month}-${day < 10
        ? `0${day}`
        : day}`;
    }
  }

  render() {
    const { question } = this.props;
    const { dateValue, timeValue } = this;

    return (
      <div className="txt">
        <label className="txt-l" htmlFor={question.code}>
          {question.description} {maybeRenderRequired(question.required)}
        </label>

        <div className={INPUT_MESSAGE_CONTAINER_STYLE}>
          <input
            placeholder="mm/dd/yyyy"
            name={question.code}
            id={question.code}
            className={`txt-f ${INPUT_STYLE.toString()} ${this.isDateValid ||
            this.dateFocus
              ? ''
              : 'txt-f--err'}`}
            value={dateValue}
            onChange={this.handleDateChange}
            onFocus={this.handleDateFocus}
            onBlur={this.handleDateBlur}
            aria-required={question.required}
          />

          {this.showTime &&
            <input
              placeholder="hh:mm am/pm"
              name={`${question.code} time`}
              className={`txt-f ${INPUT_STYLE.toString()} ${this.isTimeValid ||
              this.timeFocus
                ? ''
                : 'txt-f--err'}`}
              value={timeValue}
              onChange={this.handleTimeChange}
              onFocus={this.handleTimeFocus}
              onBlur={this.handleTimeBlur}
              aria-required={question.required}
              aria-label={`${question.description} time field`}
            />}

          <span className="t--err">
            {!this.isDateValid &&
              !this.dateFocus &&
              'Please use mm/dd/yyyy to format your date. '}
            {!this.isTimeValid &&
              !this.dateFocus &&
              !this.timeFocus &&
              'Please use hh:mm am/pm to format your time. '}
          </span>
        </div>
      </div>
    );
  }
}
