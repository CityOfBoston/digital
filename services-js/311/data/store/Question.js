// @flow

import { observable, computed } from 'mobx';
import type { ObservableArray } from 'mobx';

import type { ServiceAttributeDatatype } from '../dao/graphql/types';
import type { ServiceAttribute } from '../types';

import ConditionGroup from './ConditionGroup';

type ValuesArray = Array<{|
  key: string,
  name: string,
|}>;

type ConditionalValuesArray = Array<{|
  conditions: ConditionGroup,
  values: ValuesArray,
|}>;

export default class Question {
  code: string;
  description: string;
  required: boolean;
  type: ServiceAttributeDatatype;
  malformed: boolean;
  _values: ?ValuesArray;
  _conditionalValues: ?ConditionalValuesArray;
  _dependencies: ?ConditionGroup;
  @observable value: ?string | ObservableArray<string>;

  static buildQuestions(attributes: ServiceAttribute[]): Question[] {
    const questionMap = {};
    return attributes.map(attr => {
      const question = new Question(attr, questionMap);
      questionMap[question.code] = question;
      return question;
    });
  }

  constructor(
    attribute: ServiceAttribute,
    questionMap: { [code: string]: Question } = {},
  ) {
    this.code = attribute.code;
    this.description = attribute.description;
    this.required = attribute.required;
    this.malformed = false;
    this._values = attribute.values;
    this._conditionalValues = (attribute.conditionalValues || [])
      .map(({ dependentOn, values }) => ({
        conditions: new ConditionGroup(dependentOn, questionMap),
        values,
      }));
    this._dependencies = attribute.dependencies
      ? new ConditionGroup(attribute.dependencies, questionMap)
      : null;
    this.type = attribute.type;

    switch (this.type) {
      case 'SINGLEVALUELIST':
        this.value = null;
        break;

      case 'MULTIVALUELIST':
        this.value = [];
        break;

      default:
        this.value = '';
        break;
    }
  }

  @computed
  get visible(): boolean {
    return !this._dependencies || this._dependencies.holds;
  }

  @computed
  get requirementsMet(): boolean {
    if (this.malformed) {
      return false;
    }

    const { validatedValue } = this;
    return Array.isArray(validatedValue)
      ? !!validatedValue.length
      : !!validatedValue;
  }

  @computed
  get valueOptions(): ?ValuesArray {
    const { _values: values, _conditionalValues: conditionalValues } = this;

    if (!values || !conditionalValues) {
      return null;
    }

    const activeValuesArr = conditionalValues
      .filter(({ conditions }) => conditions.holds)
      .map(({ values: activeValues }) => activeValues);
    return [...values, ...[].concat(...activeValuesArr)];
  }

  @computed
  get validatedValue(): ?string | string[] {
    if (!this.visible) {
      return null;
    }

    const value = this.value ? this.value.slice() : null;
    const valueOptions = this.valueOptions;

    if (!valueOptions) {
      return value;
    } else if (Array.isArray(value)) {
      return value.filter(v => !!valueOptions.find(({ key }) => key === v));
    } else {
      return valueOptions.find(({ key }) => key === value) ? value : null;
    }
  }
}
