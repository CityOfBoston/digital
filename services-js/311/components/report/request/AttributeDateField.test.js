// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import Question from '../../../data/store/Question';
import AttributeDateField from './AttributeDateField';

describe('rendering', () => {
  it('renders date', () => {
    const question = new Question({
      code: 'DATE-SEEN',
      type: 'DATE',
      description: '',
      required: true,
      values: null,
      conditionalValues: null,
      dependencies: null,
    });

    question.value = '2017-12-31';

    const component = renderer.create(
      <AttributeDateField question={question} />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders time', () => {
    const question = new Question({
      code: 'DATE-SEEN',
      type: 'DATETIME',
      description: '',
      required: true,
      values: null,
      conditionalValues: null,
      dependencies: null,
    });

    // we don't set the time because of timezone issues on the testing environment

    const component = renderer.create(
      <AttributeDateField question={question} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
