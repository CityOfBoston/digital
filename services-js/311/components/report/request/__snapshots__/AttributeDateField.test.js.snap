// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders date 1`] = `
<label
  className="txt"
>
  <span
    className="txt-l"
  >
    
     
    <span
      className="t--req"
    >
      Required
    </span>
  </span>
  <div
    className={
      Object {
        "data-css-19lcmdn": "",
      }
    }
  >
    <input
      className="txt-f css-1b2781b "
      name="DATE-SEEN"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="mm/dd/yyyy"
      value="12/31/2017"
    />
    <span
      className="t--err p-a300"
    />
  </div>
</label>
`;

exports[`rendering renders time 1`] = `
<label
  className="txt"
>
  <span
    className="txt-l"
  >
    
     
    <span
      className="t--req"
    >
      Required
    </span>
  </span>
  <div
    className={
      Object {
        "data-css-19lcmdn": "",
      }
    }
  >
    <input
      className="txt-f css-1b2781b "
      name="DATE-SEEN"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="mm/dd/yyyy"
      value=""
    />
    <input
      className="txt-f css-1b2781b "
      name="DATE-SEEN time"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="hh:mm am/pm"
      value=""
    />
    <span
      className="t--err p-a300"
    />
  </div>
</label>
`;
